README
NAME
    RDF::Query::Client - get data from W3C SPARQL Protocol 1.0 servers

SYNOPSIS
     use RDF::Query::Client;
 
     my $query = RDF::Query::Client
                   ->new('SELECT DISTINCT ?s WHERE { ?s ?p ?o . }');
 
     my $iterator = $query->execute('http://example.com/sparql');
 
     while (my $row = $iterator->next) {
        print $row->{s}->as_string;
     }

DESCRIPTION
  Constructor
    "new ( $sparql, \%opts )"
        Returns a new RDF::Query::Client object for the specified $sparql.
        The object's interface is designed to be roughly compatible with
        RDF::Query objects, though RDF::Query is not required by this
        module.

        Options include:

        UserAgent - an LWP::UserAgent to handle HTTP requests.

        Unlike RDF::Query, where you get a choice of query language, the
        query language for RDF::Query::Client is always 'sparql'.
        RDF::TrineShortcuts offers a way to perform RDQL queries on remote
        SPARQL stores though (by transforming RDQL to SPARQL).

  Public Methods
    "execute ( $endpoint, \%opts )"
        $endpoint is a URI object or string containing the endpoint URI to
        be queried.

        Options include:

        *   UserAgent - an LWP::UserAgent to handle HTTP requests.

        *   QueryMethod - 'GET', 'POST', 'PATCH' or undef (automatic).

        *   QueryParameter - defaults to 'query'.

        *   AuthUsername - HTTP Basic authorization.

        *   AuthPassword - HTTP Basic authorization.

        *   Headers - additional headers to include (hashref).

        *   Parameters - additional GET/POST fields to include (hashref).

        *   ContentType - 'application/sparql-query',
            'application/sparql-update' or
            'application/x-www-form-urlencoded' (default).

        Returns undef on error; an RDF::Trine::Iterator if called in a
        scalar context; an array obtained by calling "get_all" on the
        iterator if called in list context.

    "discover_execute( $resource_uri, \%opts )"
        Experimental feature. Discovers a SPARQL endpoint relevant to
        $resource_uri and then calls "$query->execute" against that. Uses an
        LRDD-like method to discover the endpoint. If you're publishing data
        and want people to be able to find your SPARQL endpoint
        automatically, the easiest way is to include an Link header in HTTP
        responses:

         Link: </my/endpoint>; rel="http://ontologi.es/sparql#endpoint"

        Change the URL in the angled brackets, but not the URL in the rel
        string.

        This feature requires the HTTP::LRDD package to be installed.

    "get ( $endpoint, \%opts )"
        Executes the query using the specified endpoint, and returns the
        first matching row as a LIST of values. Takes the same arguments as
        "execute".

    "as_sparql"
        Returns the query as a string in the SPARQL syntax.

    "useragent"
        Returns the LWP::UserAgent object used for retrieving web content.

    "http_response"
        Returns the last HTTP Response the client experienced.

    "error"
        Returns the last error the client experienced.

  Security
    The "execute" and "get" methods allow AuthUsername and AuthPassword
    options to be passed to them for HTTP Basic authentication. For more
    complicated authentication (Digest, OAuth, Windows, etc), it is also
    possible to pass these methods a customised LWP::UserAgent.

    If you have the Crypt::SSLeay package installed, requests to HTTPS
    endpoints should work. It's possible to specify a client X.509
    certificate (e.g. for WebID authentication) by setting particular
    environment variables. See Crypt::SSLeay documentation for details.

BUGS
    Probably.

    Please report any you find here:
    <https://rt.cpan.org/Dist/Display.html?Queue=RDF-Query-Client>.

SEE ALSO
    *   RDF::Trine, RDF::Trine::Iterator

    *   RDF::Query

    *   LWP::UserAgent

    *   <http://www.w3.org/TR/rdf-sparql-protocol/>

    *   <http://www.w3.org/TR/rdf-sparql-query/>

    *   <http://www.perlrdf.org/>

AUTHOR
    Toby Inkster, <tobyink@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2009-2013 by Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
RDF-Query-Client
================

Created:      2009-11-10
Home page:    <https://metacpan.org/release/RDF-Query-Client>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=RDF-Query-Client>
Maintainer:   Toby Inkster <tobyink@cpan.org>

0.113	2014-09-10

 [ Packaging ]
 - Switch to Dist::Inkt.

0.112	2013-07-16

 - Prefer SPARQL results in XML over JSON. (This partially rolls back a
   change from 0.104.)

0.111	2013-07-02

 [ Bug Fixes ]
 - Stop bundling old version of Scalar::Util in 'inc'.

0.110	2012-03-13

 [ Bug Fixes ]
 - sparql.org changed some URIs, breaking our tests.
   Fixes RT#75676.
   <https://rt.cpan.org/Ticket/Display.html?id=75676>

0.109	2012-02-08

 [ Bug Fixes ]


0.108	2012-02-08

 - Updated: Stop using deprecated UNIVERSAL::AUTHORITY.
   Jonas Smedegaard++

0.107	2011-12-03

 [ Bug Fixes ]
 - Update to newer Module::Package::RDF

0.106	2011-11-26

 [ Packaging ]
 - Module::Package::RDF

 [ Other ]
 - General tidy-up.
 - Support 'Content-Type: application/sparql-(query|update)'.
 - Support HTTP 'PATCH' method.
 - Updated: use 5.010

0.105	2011-03-04

 - Fix annoying warning message.

0.104	2011-02-23

 - Accept (and prefer) JSON results format.
 - Clean up media-type handling.
 - Delegate media-type-to-parser mapping to RDF::Trine.

0.103	2010-07-08

 [ Bug Fixes ]
 - use Scalar::Util qw'blessed'

0.102	2010-07-07

 [ Bug Fixes ]
 - Construct HTTP::LRDD object just in time.

0.101	2010-07-03

 [ Bug Fixes ]
 - Possibly this will fix some cpan-tester fails, though I'm not entirely
   sure.

0.100	2010-06-30

 - Added: Automatic endpoint discovery.

0.05	2010-03-20

 - Minor documentation improvements.
 - Updated: Test script now checks HTTP connectivity to sparql.org before
   using it for tests.

0.04	2010-01-27

 - Updated: Use my new build system.

0.03	2009-12-07

 - Makefile.PL compatibility with earlier Perl versions.
 - Remove documentation for RDF::Query-compatible stub functions; issue
   warnings when they're called.

0.02	2009-11-12

 - Add security notes to documentation
 - Added: Implement get() method
 - Added: Support for HTTP Basic Auth
 - lib/RDF/Query/Client.pm: use 5.006

0.01	2009-11-10


lib/RDF/Query/Client.pm
package RDF::Query::Client;

use 5.010;
use strict;
use warnings;

use Carp 0 qw/carp/;
use LWP::UserAgent 0 qw//;
use RDF::Trine 0.133 qw//;
use Scalar::Util 0 qw/blessed/;
use URI::Escape 0 qw/uri_escape/;

BEGIN {
	$RDF::Query::Client::AUTHORITY = 'cpan:TOBYINK';
	$RDF::Query::Client::VERSION   = '0.108';
}

use namespace::clean;
use authority::shared q<http://www.perlrdf.org/>;

sub new
{
	my ($class, $query, $opts) = @_;
	
	my $self = bless {
		query      => $query ,
		useragent  => ($opts->{UserAgent} // undef) ,
		results    => [] ,
		error      => undef ,
		}, $class;
	
	return $self;
}

sub execute
{
	my ($self, $endpoint, $opts) = @_;
	
	my $ua       = $opts->{UserAgent} // $self->useragent;
	my $request  = $self->_prepare_request($endpoint, $opts);
	my $response = $ua->request($request);
	
	push @{ $self->{results} }, { response => $response };
	
	my $iterator = $self->_create_iterator($response);
	return unless defined $iterator;
	$self->{results}[-1]{iterator} = $iterator;
	
	if (wantarray)
	{
		return $iterator->get_all;
	}
	else
	{
		return $iterator;
	}
}

our $LRDD;
sub discover_execute
{
	my ($self, $resource_uri, $opts) = @_;
	
	unless ($LRDD)
	{
		no warnings;
		eval 'use HTTP::LRDD;';
		eval
		{
			$LRDD = HTTP::LRDD->new('http://ontologi.es/sparql#endpoint', 'http://ontologi.es/sparql#fingerpoint')
				if HTTP::LRDD->can('new');
		};
	}
	
	unless (blessed($LRDD) and $LRDD->isa('HTTP::LRDD'))
	{
		$self->{error} = "Need HTTP::LRDD to use the discover_execute feature.";
		return;
	}
	
	my $endpoint = $LRDD->discover($resource_uri)
		or return;
	
	return $self->execute($endpoint, $opts);
}

sub get
{
	my $stream = execute(@_);
	
	if (ref $stream)
	{
		if ($stream->is_bindings)
		{
			my $row = $stream->next;
			return $stream->binding_values;
		}
		if ($stream->is_graph)
		{
			my $st = $stream->next;
			return ($st->subject, $st->predicate, $st->object);
		}
		if ($stream->is_boolean)
		{
			my @rv;
			push @rv, 1 if $stream->get_boolean;
			return @rv;
		}
	}
	
	return;
}

sub as_sparql
{
	return (shift)->{query};
}

sub http_response
{
	return (shift)->{results}[-1]{response};
}

sub error
{
	return (shift)->{error};
}

sub prepare { carp "Method not implemented"; }
sub execute_plan { carp "Method not implemented"; }
sub execute_with_named_graphs { carp "Method not implemented"; }
sub aggregate { carp "Method not implemented"; }
sub pattern { carp "Method not implemented"; }
sub sse { carp "Method not implemented"; }
sub algebra_fixup { carp "Method not implemented"; }
sub add_function { carp "Method not implemented"; }
sub supported_extensions { carp "Method not implemented"; }
sub supported_functions { carp "Method not implemented"; }
sub add_computed_statement_generator { carp "Method not implemented"; }
sub get_computed_statement_generators { carp "Method not implemented"; }
sub net_filter_function { carp "Method not implemented"; }
sub add_hook_once { carp "Method not implemented"; }
sub add_hook { carp "Method not implemented"; }
sub parsed { carp "Method not implemented"; }
sub bridge { carp "Method not implemented"; }
sub log { carp "Method not implemented"; }
sub logger { carp "Method not implemented"; }
sub costmodel { carp "Method not implemented"; }

sub useragent
{
	my ($self) = @_;
	
	unless (defined $self->{useragent})
	{
		my $accept = join q{, },
			'application/sparql-results+json',
			'application/sparql-results+xml;q=0.9',
			'application/rdf+xml',
			'application/x-turtle',
			'text/turtle';
		my $agent  = sprintf('%s/%s (%s) ',
			__PACKAGE__,
			__PACKAGE__->VERSION,
			__PACKAGE__->AUTHORITY,
			);
		$self->{useragent} = LWP::UserAgent->new(
			agent             => $agent,
			max_redirect      => 2,
			parse_head        => 0,
			protocols_allowed => [qw/http https/],
			);
		$self->{useragent}->default_header(Accept => $accept);
	}
	
	$self->{useragent};
}

sub _prepare_request
{
	my ($self, $endpoint, $opts) = @_;
	
	my $method = uc ($opts->{QueryMethod} // '');
	if ($method !~ /^(get|post|patch)$/i)
	{
		$method = (length $self->{'query'} > 511) ? 'POST' : 'GET';
	}
	
	my $param = $opts->{QueryParameter} // 'query';
	
	my $uri = '';
	my $cnt = '';
	if ($method eq 'GET')
	{
		$uri  = $endpoint . ($endpoint =~ /\?/ ? '&' : '?');
		$uri .= sprintf(
			"%s=%s",
			uri_escape($param),
			uri_escape($self->{query})
			);
		if ($opts->{Parameters})
		{
			foreach my $field (keys %{$opts->{Parameters}})
			{
				$uri .= sprintf(
					"&%s=%s",
					uri_escape($field),
					uri_escape($opts->{Parameters}->{$field}),
					);
			}
		}
	}
	elsif ($method eq 'POST')
	{
		$uri  = $endpoint;
		$cnt  = sprintf(
			"%s=%s",
			uri_escape($param),
			uri_escape($self->{query})
			);
		if ($opts->{Parameters})
		{
			foreach my $field (keys %{$opts->{Parameters}})
			{
				$cnt .= sprintf(
					"&%s=%s",
					uri_escape($field),
					uri_escape($opts->{Parameters}{$field}),
					);
			}
		}
	}
	
	my $req = HTTP::Request->new($method => $uri);
	
	my $type = $opts->{ContentType} // '';
	if ($type =~ m{^application/sparql-query}i)
	{
		$req->content_type('application/sparql-query');
		$req->content($self->{query});
	}
	elsif ($type =~ m{^application/sparql-update}i)
	{
		$req->content_type('application/sparql-update');
		$req->content($self->{query});
	}
	else
	{
		$req->content_type('application/x-www-form-urlencoded');
		$req->content($cnt);
	}
	
	$req->authorization_basic($opts->{AuthUsername}, $opts->{AuthPassword})
		if defined $opts->{AuthUsername};
	foreach my $k (keys %{$opts->{Headers}})
	{
		$req->header($k => $opts->{Headers}{$k});
	}
	
	return $req;
}

sub _create_iterator
{
	my ($self, $response) = @_;
	
	unless ($response->is_success)
	{
		$self->{error} = $response->message;
		return;
	}
	
	if ($response->content_type eq 'application/sparql-results+xml'
	or  $response->content_type eq 'application/sparql-results+json')
	{
		local $@ = undef;
		my $iterator = eval
		{
			if ($response->content_type eq 'application/sparql-results+json')
				{ RDF::Trine::Iterator->from_json($response->decoded_content); }
			else
				{ RDF::Trine::Iterator->from_string($response->decoded_content); }
		};
		return $iterator
			if $iterator;
		
		$self->{error} = $@;
		return;
	}
	else
	{
		my $model;
		eval
		{
			my $parser = RDF::Trine::Parser->parser_by_media_type($response->content_type);
			my $tmp    = RDF::Trine::Model->temporary_model;
			$parser->parse_into_model($response->base, $response->decoded_content, $tmp);
			$model = $tmp;
		};
		
		return $model->as_stream if defined $model;
		
		$self->{error} = sprintf("Response of type '%s' could not be parsed.", $response->content_type);
		return;
	}
}

1;

__END__

=head1 NAME

RDF::Query::Client - get data from W3C SPARQL Protocol 1.0 servers

=head1 SYNOPSIS

 use RDF::Query::Client;
 
 my $query = RDF::Query::Client
               ->new('SELECT DISTINCT ?s WHERE { ?s ?p ?o . }');
 
 my $iterator = $query->execute('http://example.com/sparql');
 
 while (my $row = $iterator->next) {
    print $row->{s}->as_string;
 }

=head1 DESCRIPTION

=head2 Constructor

=over 4

=item C<< new ( $sparql, \%opts ) >>

Returns a new RDF::Query::Client object for the specified C<$sparql>.
The object's interface is designed to be roughly compatible with RDF::Query
objects, though RDF::Query is not required by this module.

Options include:

=over 4

=item B<UserAgent> - an LWP::UserAgent to handle HTTP requests.

=back 

Unlike RDF::Query, where you get a choice of query language, the query
language for RDF::Query::Client is always 'sparql'. RDF::TrineShortcuts offers
a way to perform RDQL queries on remote SPARQL stores though (by transforming
RDQL to SPARQL).

=back

=head2 Public Methods

=over 4

=item C<< execute ( $endpoint, \%opts ) >>

C<$endpoint> is a URI object or string containing the endpoint
URI to be queried.

Options include:

=over 4

=item * B<UserAgent> - an LWP::UserAgent to handle HTTP requests.

=item * B<QueryMethod> - 'GET', 'POST', 'PATCH' or undef (automatic).

=item * B<QueryParameter> - defaults to 'query'.

=item * B<AuthUsername> - HTTP Basic authorization.

=item * B<AuthPassword> - HTTP Basic authorization.

=item * B<Headers> - additional headers to include (hashref).

=item * B<Parameters> - additional GET/POST fields to include (hashref).

=item * B<ContentType> - 'application/sparql-query',
'application/sparql-update' or 'application/x-www-form-urlencoded' (default).

=back

Returns undef on error; an RDF::Trine::Iterator if called in a
scalar context; an array obtained by calling C<get_all> on the
iterator if called in list context.

=item C<< discover_execute( $resource_uri, \%opts ) >>

Experimental feature. Discovers a SPARQL endpoint relevent to $resource_uri
and then calls C<< $query->execute >> against that. Uses an LRDD-like
method to discover the endpoint. If you're publishing data and want people
to be able to find your SPARQL endpoint automatically, the easiest way is to
include an Link header in HTTP responses:

 Link: </my/endpoint>; rel="http://ontologi.es/sparql#endpoint"

Change the URL in the angled brackets, but not the URL in the rel string.

This feature requires the HTTP::LRDD package to be installed.

=item C<< get ( $endpoint, \%opts ) >>

Executes the query using the specified endpoint, and returns the first
matching row as a LIST of values. Takes the same arguments as C<execute>.

=item C<< as_sparql >>

Returns the query as a string in the SPARQL syntax.

=item C<< useragent >>

Returns the LWP::UserAgent object used for retrieving web content.

=item C<< http_response >>

Returns the last HTTP Response the client experienced.

=item C<< error >>

Returns the last error the client experienced.

=back

=head2 Security

The C<execute> and C<get> methods allow AuthUsername and
AuthPassword options to be passed to them for HTTP Basic authentication.
For more complicated authentication (Digest, OAuth, Windows, etc),
it is also possible to pass these methods a customised LWP::UserAgent.

If you have the Crypt::SSLeay package installed, requests to HTTPS
endpoints should work. It's possible to specify a client X.509
certificate (e.g. for WebID authentication) by setting particular
environment variables. See L<Crypt::SSLeay> documentation for details.

=head1 BUGS

Probably.

Please report any you find here:
L<https://rt.cpan.org/Dist/Display.html?Queue=RDF-Query-Client>.

=head1 SEE ALSO

=over 4

=item * L<RDF::Trine>, L<RDF::Trine::Iterator>

=item * L<RDF::Query>

=item * L<LWP::UserAgent>

=item * http://www.w3.org/TR/rdf-sparql-protocol/

=item * http://www.w3.org/TR/rdf-sparql-query/

=item * http://www.perlrdf.org/

=back

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009-2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=cut



Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "get data from W3C SPARQL Protocol 1.0 servers",
  "author"         => [
                        "Toby Inkster <tobyink\@cpan.org>",
                        "TOBYINK <tobyink\@cpan.org>",
                      ],
  "dynamic_config" => 0,
  "generated_by"   => "Moose::Meta::Class::__ANON__::SERIAL::11 version , CPAN::Meta::Converter version 2.140640",
  "keywords"       => ["HTTP", "Protocol", "Query", "RDF", "REST", "SPARQL"],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "RDF-Query-Client",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          recommends => { "HTTP::LRDD" => 0 },
                          requires   => {
                                          "authority::shared" => 0,
                                          "Carp"              => 0,
                                          "HTTP::Request"     => 0,
                                          "LWP::UserAgent"    => 0,
                                          "namespace::clean"  => "0.20",
                                          "RDF::Trine"        => 0.133,
                                          "Scalar::Util"      => 0,
                                          "URI::Escape"       => 0,
                                        },
                        },
                        test => { requires => { "Test::More" => 0.61 } },
                      },
  "provides"       => {
                        "RDF::Query::Client" => { file => "lib/RDF/Query/Client.pm", version => 0.113 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=RDF-Query-Client",
                                        },
                        homepage     => "https://metacpan.org/release/RDF-Query-Client",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-rdf-query-client.git",
                                          web  => "https://github.com/tobyink/p5-rdf-query-client",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/RDF-Query-Client/project",
                      },
  "version"        => 0.113,
  "x_contributors" => ["Jonas Smedegaard", "TOBYINK"],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

